### ISSUE TITLE
M5 Max 128GB results: val_bpb 1.574 → 1.235 (25 experiments)

### ISSUE THREAD

**Issue by hirokimsd** [trevin-creator/autoresearch-mlx#21]
## Summary

- **Hardware**: Apple M5 Max, 128GB unified memory
- **Best val_bpb**: **1.235** (21.5% improvement over baseline 1.574)
- **25 experiments** conducted autonomously (14 keep, 8 discard, 3 crash)
- **Agent**: Claude Code (Opus 4.6) in interactive mode via tmux

## Key findings on M5 Max

- **Smaller batch (2^15)** is the biggest single lever, but 2^14 was too small — sweet spot exists
- **Simplification wins**: removing x0 connection, removing logit cap, cleaning unused params all improved val_bpb
- **SiLU was catastrophic** (1.301 → 1.768) — ReLU² is clearly better for this architecture
- **Gradual LR reduction** was the most reliable optimization path (0.04 → 0.02 → 0.01 → 0.008 → 0.006 → 0.005, each step improved)
- **VE on all layers** beats alternating pattern
- **Depth 6 crashed** under memory pressure; 4 layers is the sweet spot for 5-min budget

## Final configuration

```
Architecture: 4 layers, n_embd=256, 2 heads, 3x MLP, ReLU², QK-norm
             SSSL window pattern, VE on all layers, no x0 connection
Optimizer:   AdamW, matrix_lr=0.005, embedding_lr=0.6, weight_decay=0.15
             betas=(0.8, 0.95), eps=1e-8
Schedule:    warmup=0.1, warmdown=0.4, final_lr_frac=0.0
Training:    batch_size=2^15, ~1900 steps in 5 min, ~120K tok/sec
```

## Results

| val_bpb | Status | Description |
|---:|---|---|
| 1.574 | keep | baseline (4-layer, default config) |
| 1.428 | keep | halve batch size to 2^15 |
| 1.413 | keep | remove logit cap (tanh capping) |
| 1.392 | keep | 3x MLP ratio instead of 4x |
| 1.350 | keep | lower matrix LR 0.04 → 0.02 |
| 1.326 | keep | increase weight decay 0.2 → 0.3 |
| 1.301 | keep | add 10% warmup |
| 1.303 | discard | extend warmdown 0.5 → 0.6 |
| 1.353 | discard | halve batch to 2^14 (too small) |
| 1.768 | discard | SiLU (much worse) |
| crash | crash | depth 6 (memory pressure) |
| 1.284 | keep | matrix LR → 0.01 |
| 1.269 | keep | remove x0 connection (simplification win) |
| 1.263 | keep | matrix LR → 0.008 |
| 1.254 | keep | matrix LR → 0.006 |
| 1.252 | keep | weight decay 0.3 → 0.2 |
| 1.246 | keep | VE on all layers |
| 1.242 | keep | matrix LR → 0.005 |
| 1.241 | keep | warmdown 0.5 → 0.4 |
| 1.240 | keep | weight decay 0.2 → 0.15 |
| 1.242 | keep | clean up unused params (simplification) |
| **1.235** | **keep** | **adam eps 1e-8** |

## Comparison with other hardware

| Machine | Best val_bpb | Starting point |
|---|---:|---:|
| **M5 Max 128GB** | **1.235** | **1.574** |
| M4 Max (README) | 1.295 | 1.597 |
| Mac Mini (README) | 1.353 | 1.922 |

## Interesting contrasts with other machines

The M5 Max found a different recipe than what's reported for M4 Max and Mac Mini:
- **SiLU**: beneficial on M4 Max #2, catastrophic on M5 Max
- **Muon optimizer**: won on Mac Mini, AdamW won on M5 Max
- This reinforces the README's observation that hardware-specific behavior is real

Full git history available on branch `autoresearch/m5max-apr5`.

Thanks for the great project — the "simplicity criterion" in program.md was validated beautifully by the experiments. 🙏
